new extension for contributing to +Add page

christianvogt opened the pull request:
**Fixes**: 
https://issues.redhat.com/browse/CNV-4395

**Analysis / Root cause**: 
The +Add page in the developer perspective is hard coded to show a fixed set of tiles. There are 2 contributing plugins and soon to be a third:
- dev-console itself
- knative-plugin
- kubevirt (future)

This page should be extensible so that dependencies flow in the right direction and contributions are decoupled from dev-console.

**Solution Description**: 
Added a new extension point that allows tile contributions to the +Add page.
Updated knative-plugin to contribute event source tile through extensions.
Updated dev-console to contribute all its tiles through extensions.

**Screen shots / Gifs for design review**: 
No visual changes.

**Browser conformance**: 
<!-- To mark tested browsers, use [x] -->
- [ ] Chrome
- [x] Firefox
- [x] Safari
- [x] Edge

jelkosz commented:
@christianvogt Thank you!

mareklibra commented:
/lgtm
works nice!

vojtechszocs commented on frontend/packages/dev-console/src/components/EmptyState.tsx:
```diff
@@ -2,41 +2,67 @@ import * as React from 'react';
+}) => {
+  const access =
+    !accessReview ||
+    (Array.isArray(accessReview) ? accessReview : [accessReview])
```
I'd suggest to declare `accessReview` simply as:

```ts
accessReview?: AccessReviewResourceAttributes[];
```

If that's not feasible, we can use `_.castArray` to simplify the coercion to array object:

```ts
_.castArray(accessReview).map( /* .. */ );
```

christianvogt replied:
Converted to array property.

vojtechszocs commented on frontend/packages/dev-console/src/extensions/add-actions.ts:
```diff
@@ -0,0 +1,29 @@
+namespace ExtensionProperties {
+  export type AddAction = {
+    /** Optional ID used to identify the action. */
+    id?: string;
```
I'd suggest making `id` mandatory for both safety and consistency (assuming there can be multiple extension instances with the same `label`).

vojtechszocs replied:
Also, we can add `data-test-id={id}` to `CatalogTile` component to preserve existing behavior.

vojtechszocs commented on frontend/packages/dev-console/src/extensions/add-actions.ts:
```diff
@@ -0,0 +1,29 @@
+  export type AddAction = {
+    /** Optional ID used to identify the action. */
+    id?: string;
+    /** The label of the actio.n */
```
```suggestion
    /** The label of the action. */
```

vojtechszocs commented on frontend/packages/dev-console/src/extensions/add-actions.ts:
```diff
@@ -0,0 +1,29 @@
+    id?: string;
+    /** The label of the actio.n */
+    label: string;
+    /** The description of the action.  */
```
```suggestion
    /** The description of the action. */
```

vojtechszocs commented on frontend/packages/dev-console/src/extensions/add-actions.ts:
```diff
@@ -0,0 +1,29 @@
+    /** The URL to navigate to. */
+    url: string;
+    /** Optional access review to control visibility / enablement of the action. */
+    accessReview?: AccessReviewResourceAttributes | AccessReviewResourceAttributes[];
```
Based on my suggestion above:

```suggestion
    accessReview?: AccessReviewResourceAttributes[];
```

Type simplicity reflects into simplicity of code that interprets the values.

christianvogt replied:
I'll simplify the type to just an array.

christianvogt commented:
/hold to address comments

christianvogt commented:
/hold cancel

christianvogt commented:
@vojtechszocs addressed your comments.

@mareklibra I incorporated your changes from https://github.com/openshift/console/pull/4940 because it actually helped me simplify one use case. Once this merges you can close your PR. I also used the updated regex.

vojtechszocs commented:
Thanks for addressing my review comments.

/lgtm

vojtechszocs commented:
/retest

christianvogt commented:
/test e2e-gcp-console
